Caching severly degraded in 8.0

### Summary

## Brief
This change https://github.com/concourse/concourse/pull/9330 has caused caching to be essentially useless for build steps that utilise CMake.

### Detail
Some build systems - for some reason unbeknownst to me - encode absolute paths in generated intermediary build files. CMake definitely does this. This means that, if you mark the build-tree directory as a cache in a build step, for a project using CMake, this build-tree is useless as a cache. In fact, it's not only useless - which would be annoying enough - but it actually causes builds to fail (names altered);

```
selected worker: builder-01

HEAD is now at 07b644e Removing usage of singleton creation of functors

From ssh://gerrit/some-project

 * branch            07b644e3cf94756b5a363578902cdd60a3a6fbf9 -> FETCH_HEAD

HEAD is now at 07b644e Removing usage of singleton creation of functors

CMake Error: The source directory "/tmp/build/c91a22c6/src" does not exist.

Specify --help for usage, or press the help button on the CMake GUI.

FAILED: build.ninja 

/usr/bin/cmake --regenerate-during-build -S/tmp/build/c91a22c6/src -B/tmp/build/c91a22c6/build

ninja: error: rebuilding 'build.ninja': subcommand failed

Error:  Build failure
```

I'm aware that the above actually shows that CMake doesn't like the _source_ directory changing path, not the build tree. However, I promise you that if you symlink the source directory such that CMake believes it to be in the same place, it fails just as quickly.

Hijacking the build step yields:

```
% fly hijack -s build -j linux/code-review-unit-test
1: build #190.3, step: build, type: task
2: build #190.4, step: build, type: task
choose a container: 2
root@22b17fec-913b-4d70-81ee-1dc99c622017:/tmp/build/75f6c480# pwd
/tmp/build/75f6c480
```

This path has changed because https://github.com/concourse/concourse/pull/9330 has caused the random prefix to be generated from not just the name - an unchanging value - but also the current time.

Note, I cannot control this behaviour at all. See https://discourse.cmake.org/t/change-directory-of-build/8671

As an aside, I'm eager to understand what problem this CR was intending to solve - the current working directory isn't something you can exactly hide from the user, well, sans namespacing, chrooting, and, well.. i guess that's what running inside of a container means :) although I suppose that begs the question of why bother randomising the prefix..! Actually, thinking about it, the fact that there is a random prefix _at all_ doesn't feel very cash-money. No pun intended.

More seriously, this _absolutely_ will break all CMake based builds, if caching is used.

I'm going to locally revert this change and report back the results, but I will be surprised if this doesn't resolve the issue.

Thanks,

### Steps to Reproduce

1. Create a task that generates a build-tree from a project using CMake.
2. Use that build tree as a cache.
3. Run one build, then another. (Either as a "re-run with same inputs", or with new inputs/triggers)

### Expected Results

I expected the cache to cache the build tree, and for it to be useable as a cache.

### Actual Results

The build tree was cachced, but given a different path. Upon re-runing the build from this build-tree, CMake violently disapproves and fails.

### Web Node(s) configuration

One web node. Irrelevant.

### Worker(s) configuration

(no redactions required)
```
CONCOURSE_WORK_DIR="/var/concourse/builds"
CONCOURSE_TSA_HOST="172.16.1.191:2222"
CONCOURSE_TSA_PUBLIC_KEY="/var/concourse/etc/concourse/keys/tsa_host_key.pub"
CONCOURSE_TSA_WORKER_PRIVATE_KEY="/var/concourse/etc/concourse/keys/worker_key"
CONCOURSE_RUNTIME="containerd"
CONCOURSE_CONTAINERD_DNS_SERVER="172.16.1.1"
CONCOURSE_CONTAINERD_ALLOW_HOST_ACCESS=1
CONCOURSE_WORKER_BAGGAGECLAIM_DRIVER="overlay"
```

### Concourse Version

8.0.0

### Browser (if applicable)

_No response_

### Did this use to work?

Yes, prior to https://github.com/concourse/concourse/pull/9330

--- a/atc/engine/step_factory.go
+++ b/atc/engine/step_factory.go
@@ -177,7 +177,7 @@ func (factory *coreStepFactory) TaskStep(
 	containerMetadata db.ContainerMetadata,
 	delegateFactory DelegateFactory,
 ) exec.Step {
-	sum := sha256.Sum256(fmt.Appendf([]byte{}, "%d", time.Now().Unix()))
+	sum := sha256.Sum256([]byte(plan.Task.Name))
 	containerMetadata.WorkingDirectory = filepath.Join("/tmp", "build", fmt.Sprintf("%x", sum[:4]))
 
 	taskStep := exec.NewTaskStep(

